src/Microsoft.ComponentDetection.Detectors/go/GoComponentWithReplaceDetector.cs

```diff
@@ -0,0 +1,560 @@
+
+        var generateGraphProcess = await this.commandLineInvocationService.ExecuteCommandAsync("go", null, workingDirectory: projectRootDirectory, new List<string> { "mod", "graph" }.ToArray());
+        if (generateGraphProcess.ExitCode == 0)
+        {
+            this.PopulateDependencyGraph(generateGraphProcess.StdOut, singleFileComponentRecorder);
+            record.WasGraphSuccessful = true;
+        }
+
+        return true;
+    }
+
+    private void CreateListOfGoComponents(string line, ISingleFileComponentRecorder singleFileComponentRecorder)
+    {
+        if (line.Trim().StartsWith("//"))
+        {
+            // this is a comment line, ignore it
+            return;
+        }
+
+        if (this.TryToCreateGoComponentFromModLine(line, out var goComponent))
+        {
+            this.GoComponents.Add(goComponent);
+        }
+        else
+        {
+            var lineTrim = line.Trim();
+            this.Logger.LogWarning("Line could not be parsed for component [{LineTrim}]", lineTrim);
+            singleFileComponentRecorder.RegisterPackageParseFailure(lineTrim);
+        }
+    }
+
+    private void ReplaceGoComponents(string line, ISingleFileComponentRecorder singleFileComponentRecorder)
+    {
+        if (this.TryToCreateReplacementGoComponentFromModLine(line, out var goComponent))
+        {
+            var goComponentsWithReplacementVersion = this.GoComponents.Where(component => component.Name == goComponent.Name);
+            if (goComponentsWithReplacementVersion.Any())
+            {
+                foreach (var component in goComponentsWithReplacementVersion)
+                {
+                    if (component.Name == goComponent.Name)
+                    {
+                        component.Version = goComponent.Version;
+                    }
+                }
+            }
+        }
+        else
+        {
+            var lineTrim = line.Trim();
+            this.Logger.LogWarning("Line could not be parsed for component [{LineTrim}]", lineTrim);
+            singleFileComponentRecorder.RegisterPackageParseFailure(lineTrim);
+        }
+    }
+
+    private void TryRegisterDependencyFromModLine(ISingleFileComponentRecorder singleFileComponentRecorder)
+    {
+        this.GoComponents.ForEach(goComponent => singleFileComponentRecorder.RegisterUsage(new DetectedComponent(goComponent)));
+    }
+
+    private async Task ParseGoModFileAsync(
+        ISingleFileComponentRecorder singleFileComponentRecorder,
+        IComponentStream file,
+        GoGraphTelemetryRecord goGraphTelemetryRecord)
+    {
+        using var reader = new StreamReader(file.Stream);
+
+        // There can be multiple require( ) sections in go 1.17+. loop over all of them.
+        while (!reader.EndOfStream)
+        {
+            var line = await reader.ReadLineAsync();
+
+            while (line != null && !line.StartsWith("require ("))
```

Comment:
@amitla1 - Not sure if you were aware, but I think there is no need to parse the require statements directly [go list -m](https://go.dev/ref/mod#go-list-m) does return a property called "Replace" with the module that will replace the existing entry. We are already running that command so it shouldn't add cost there, but it won't be available for the fallback scenario.